### PR TITLE
Allow mixed type value when apply to DateFilter.

### LINKED ISSUE #3858: ORM DateFilter throw fatal error when user submit unexpected query params.
**API Platform version(s) affected**: 2.5.7

**Description**  

Currently when user submit unexpected query param type like: `&date[after][][][]=1` fatal error will be throw because it's wrong type when pass it to `andWhere` method.

**How to reproduce**  

Submit query param `date[after][][][]=1`

**Possible Solution**  

Allow mixed type of `value` arg at [DateFilter#L131](https://github.com/api-platform/core/blob/master/src/Bridge/Doctrine/Orm/Filter/DateFilter.php#L131) and change exception type to `\Throwable` at [DateFilter#L136](https://github.com/api-platform/core/blob/master/src/Bridge/Doctrine/Orm/Filter/DateFilter.php#L136).

### PR DESCRIPTION
**Pull request by vuongxuongminh**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tickets       | fixes #3858 
| License       | MIT

### CODE REVIEW

File: src/Bridge/Doctrine/MongoDbOdm/Filter/DateFilter.php [hunk 1]
```diff
@@ -107,11 +107,11 @@ protected function filterProperty(string $property, $values, Builder $aggregatio
     /**
      * Adds the match stage according to the chosen null management.
      */
-    private function addMatch(Builder $aggregationBuilder, string $field, string $operator, string $value, string $nullManagement = null): void
+    private function addMatch(Builder $aggregationBuilder, string $field, string $operator, $value, string $nullManagement = null): void
     {
         try {
             $value = new \DateTime($value);
-        } catch (\Exception $e) {
+        } catch (\Throwable $e) {
```
Comment (dunglas) [inline src/Bridge/Doctrine/MongoDbOdm/Filter/DateFilter.php:114]: This should be reverted. A `\Throwable` should never be thrown. Instead, we should test that the $value is a string before passing it to the `DateTime` constructor.

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Bridge/Doctrine/Common/Filter/DateFilterTrait.php [hunk 2]
```diff
@@ -79,4 +80,20 @@ protected function getFilterDescription(string $property, string $period): array
             ],
         ];
     }
+
+    /**
+     * Normalize the value.
+     */
+    private function normalizeValue($value, string $operator): ?string
+    {
+        if (!\is_string($value)) {
```
Comment (dunglas) [inline src/Bridge/Doctrine/Common/Filter/DateFilterTrait.php:89]: Nitpicking but I would transform this test in a guard clause to simplify.

**Review by dunglas — APPROVED** [review]
LGTM!

**Review by dunglas — APPROVED** [review]

File: src/Bridge/Doctrine/Common/Filter/DateFilterTrait.php [hunk 3]
```diff
@@ -79,4 +80,20 @@ protected function getFilterDescription(string $property, string $period): array
             ],
         ];
     }
+
+    /**
+     * Normalize the value.
+     */
+    private function normalizeValue($value, string $operator): ?string
```
Comment (drupol) [inline src/Bridge/Doctrine/Common/Filter/DateFilterTrait.php:87]: I might be picky on this, but wouldn't it be better (semantically wise) to use early returns?

First exclude the bad scenario and then only return the happy scenario ?

```suggestion
    private function normalizeValue($value, string $operator): ?string
    {
            if (false === \is_string($value)) {
               $this->getLogger()->notice('Invalid filter ignored', [
                   'exception' => new InvalidArgumentException(sprintf('Invalid value for "[%s]", expected string', $operator)),
               ]);

               return null;
            }

            return $value;
    }
```

Comment (dunglas) [reply src/Bridge/Doctrine/Common/Filter/DateFilterTrait.php:87]: Using a guard clause reduces the cyclomatic complexity and makes the code more readable (imho). I would keep this block as is.

Comment (drupol) [reply src/Bridge/Doctrine/Common/Filter/DateFilterTrait.php:87]: But this is what I'm trying to say here :-)

First the guard (_the unhappy scenario_), then relevant returned `$value` (_the happy scenario_).

**Comment by vuongxuongminh** [comment]
Is there anything needed to get this merged?

File: src/Bridge/Doctrine/Common/Filter/DateFilterTrait.php (hunk 3 above)
Comment (alanpoulain) [reply src/Bridge/Doctrine/Common/Filter/DateFilterTrait.php:87]: I'm :+1: for @drupol's suggestion.

Comment (drupol) [reply src/Bridge/Doctrine/Common/Filter/DateFilterTrait.php:84]: Let's not forget this thing :-)

Comment (drupol) [reply src/Bridge/Doctrine/Common/Filter/DateFilterTrait.php:84]: Oh it's done already ! Great :-)

**Review by drupol — APPROVED** [review]

**Comment by alanpoulain** [comment]
Thank you @vuongxuongminh.

**Comment by drupol** [comment]
Yes, thanks !

**Pull request merged**